snuba/subscriptions/scheduler_processing_strategy.py

```diff
@@ -257,3 +278,167 @@ def terminate(self) -> None:
+        if self.__queues:
+            tick, futures = self.__queues[0]
+            return TickSubscription(tick, futures[0], len(futures) == 1)
+        return None
+
+    def popleft(self) -> TickSubscription:
+        if self.__queues:
+            tick_message, futures = self.__queues[0]
+            subscription_future = futures.popleft()
+
+            is_empty = len(futures) == 0
+
+            if is_empty:
+                self.__queues.popleft()
+
+            should_commit = is_empty and tick_message.payload.should_commit == True
+
+            return TickSubscription(tick_message, subscription_future, should_commit)
+
+        raise IndexError()
+
+    def __len__(self) -> int:
+        return sum([len(futures) for (_tick, futures) in self.__queues])
+
+
+class ProduceScheduledSubscriptionMessage(ProcessingStrategy[CommittableTick]):
+    """"
+    This strategy is responsible for producing a message for all of the subscriptions
+    scheduled for a given tick.
+
+    The subscriptions to be scheduled are those that:
+    - Have the same partition index as the tick received, and
+    - Are scheduled to be run within the time interval defined by the tick
+    (taking the jitter into account)
+
+    The subscriptions to be scheduled are encoded and produced to the
+    scheduled topic to be picked up by the subscription executor which
+    will run the actual query later.
+
+    When a tick is submitted to this strategy, we produce all of it's
+    subscriptions and place it in a queue without waiting to see if they
+    were succesfully produced or not.
+
+    On each call to poll(), we remove completed futures from the queue.
+    If all scheduled subscription messages for that tick were succesfully
+    produced and the tick indicates it should be committed, we commit offsets.
+    """
+
+    def __init__(
+        self,
+        entity_key: EntityKey,
+        schedulers: Mapping[int, Scheduler[Subscription]],
+        producer: Producer[KafkaPayload],
+        scheduled_topic_spec: KafkaTopicSpec,
+        commit: Callable[[Mapping[Partition, Position]], None],
+    ) -> None:
+        self.__schedulers = schedulers
+        self.__encoder = SubscriptionScheduledTaskEncoder(entity_key)
+        self.__producer = producer
+        self.__scheduled_topic = Topic(scheduled_topic_spec.topic_name)
+        self.__commit = commit
+        self.__closed = False
+
+        # Stores each tick with it's futures
+        self.__queue = ScheduledSubscriptionQueue()
+
+        # Not a hard max
+        self.__max_buffer_size = 10000
+
+    def poll(self) -> None:
+        # Remove completed tasks from the queue and raise if an exception occurred.
+        # This method does not attempt to recover from any exception.
+        # Also commits any offsets required.
+        while True:
+            tick_subscription = self.__queue.peek()
+
+            if tick_subscription is None:
+                break
+
```

Comment:
```
   while self.__queue:
```
So you avoid the risk of a while true and you will have an explicit exit condition that will always be met in some way.